### PR TITLE
Feature spring cloud zuul adapter

### LINKED ISSUE #63: [RFC] Flow control for Netflix Zuul
## Issue Description

Type: *feature request*

### Describe or what feature you want

Integrate Sentinel with Netflix Zuul so that we can provide out-of-box flow control ability for Zuul (1.x and 2.x). Moreover, we may support more open-source API Gateways.

### PR DESCRIPTION
**Pull request by tigerMoon**
### Describe what this PR does / why we need it
Spring cloud zuul lack the capability of flow control. this starter implements this. 

### Does this pull request fix one issue?

Fixes #63 

### Describe how you did it
Implement `ZuulFilter` at pre, post, and error phase. provide ServiceId and API PATH level flow control.

### Describe how to verify it
1. run test cases.
2. include starter in local spring cloud project.

### Special notes for reviews
no

### CODE REVIEW

**Comment by sczyh30** [comment]
Thanks for contributing. I'll do the review these days.

**Comment by tigerMoon** [comment]
this pr has decoupled sentinel zuul adapter from Spring Cloud Zuul. the Spring Cloud Zuul adapter will put into [spring cloud alibaba](https://github.com/spring-cloud-incubator/spring-cloud-alibaba) project and dependant with this adapter.

**Comment by junjiex123** [comment]
这个adapter什么时候可以搞定？

File: sentinel-adapter/sentinel-zuul-adapter/pom.xml [hunk 1]
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>sentinel-adapter</artifactId>
+        <groupId>com.alibaba.csp</groupId>
+        <version>1.3.0-SNAPSHOT</version>
```
[the remark this answers is not in the document]
Comment (tigerMoon) [reply sentinel-adapter/sentinel-zuul-adapter/pom.xml]: should I need to git rebase this pr?

File: sentinel-adapter/sentinel-zuul-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/zuul/filters/SentinelPreFilter.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    public Object run() throws ZuulException {
+        RequestContext ctx = RequestContext.getCurrentContext();
+        String origin = parseOrigin(ctx.getRequest());
+        String serviceTarget = (String) ctx.get(SERVICE_ID_KEY);
+        // When serviceId blocked first get the service level fallback provider.
+        String fallBackRoute = serviceTarget;
+        try {
+            if (StringUtil.isNotEmpty(serviceTarget)) {
+                RecordLog.info(String.format("[Sentinel Pre Filter] Origin: %s enter ServiceId: %s", origin, serviceTarget));
+                ContextUtil.enter(serviceTarget, origin);
```
[the remark this answers is not in the document]
Comment (tigerMoon) [reply sentinel-adapter/sentinel-zuul-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/zuul/filters/SentinelPreFilter.java]:
```
zuul:
  routes:
    book:
      path: /book/**
      serviceId: book
    coke:
      path: /coke/**
      serviceId: coke
```

here book is a service, and coke is another service e.g serviceId = book.

In spring cloud zuul, 
when a route with `serviceId`, here is `serviceTarget` is not empty.
when a route with direct URL like  (http://16.27.90.19:9090/app.). then `serviceTarget` is empty.

File: sentinel-adapter/sentinel-zuul-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/zuul/filters/SentinelErrorFilter.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+            Throwable throwable = ctx.getThrowable();
+            if (throwable != null) {
+                if (!(throwable.getCause() instanceof BlockException)) {
+                    Tracer.trace(throwable.getCause());
+                    RecordLog.info("[Sentinel Error Filter] Trace cause", throwable.getCause());
+
+                }
+            }
+        } finally {
+            while (ContextUtil.getContext() != null && ContextUtil.getContext().getCurEntry() != null) {
```
[the remark this answers is not in the document]
Comment (tigerMoon) [reply sentinel-adapter/sentinel-zuul-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/zuul/filters/SentinelErrorFilter.java]: yes

**Comment by sczyh30** [comment]
Awesome, thanks for contributing!

**Comment by sczyh30** [comment]
@tigerMoon Could you please add Chinese document here: https://github.com/sentinel-group/sentinel-website/blob/master/docs/zh-cn/open-source-framework-integrations.md

We'll merge this with wiki later.

**Pull request merged**